librad/src/git/storage.rs

```diff
@@ -0,0 +1,371 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    ops::Range,
+    sync::{Arc, Mutex, MutexGuard},
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+use thiserror::Error;
+
+use crate::{
+    git::{
+        ext::{is_not_found_err, References},
+        repo::{self, Repo},
+        types::Reference,
+        url::GitUrlRef,
+    },
+    keys::SecretKey,
+    meta::entity::{
+        data::{EntityBuilder, EntityData},
+        Entity,
+    },
+    paths::Paths,
+    peer::PeerId,
+    uri::{RadUrl, RadUrn},
+};
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Blob {0} not found")]
+    NoSuchBlob(String),
+
+    #[error("Branch {0} not found")]
+    NoSuchBranch(String),
+
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+}
+
+#[derive(Clone)]
+pub struct Storage {
+    backend: Arc<Mutex<git2::Repository>>,
+    pub(crate) key: SecretKey,
+}
+
+impl Storage {
+    pub fn open(paths: &Paths, key: SecretKey) -> Result<Self, Error> {
+        git2::Repository::open_bare(paths.git_dir())
+            .map(|backend| Self {
+                backend: Arc::new(Mutex::new(backend)),
+                key,
+            })
+            .map_err(|e| e.into())
+    }
+
+    pub fn init(paths: &Paths, key: SecretKey) -> Result<Self, Error> {
+        let repo = git2::Repository::init_opts(
+            paths.git_dir(),
+            git2::RepositoryInitOptions::new()
+                .bare(true)
+                .no_reinit(true)
+                .external_template(false),
+        )?;
+
+        let mut config = repo.config()?;
+        config.set_str("user.name", "radicle")?;
+        config.set_str("user.email", &format!("radicle@{}", PeerId::from(&key)))?;
+
+        Ok(Self {
+            backend: Arc::new(Mutex::new(repo)),
+            key,
+        })
+    }
+
+    pub fn create_repo<T>(self, meta: &Entity<T>) -> Result<Repo, repo::Error>
+    where
+        T: Serialize + DeserializeOwned + Clone + Default,
+        EntityData<T>: EntityBuilder,
+    {
+        Repo::create(self, meta)
+    }
+
+    pub fn open_repo(self, urn: RadUrn) -> Result<Repo, repo::Error> {
+        Repo::open(self, urn)
+    }
+
+    pub fn clone_repo<T>(self, url: RadUrl) -> Result<Repo, repo::Error>
+    where
+        T: Serialize + DeserializeOwned + Clone + Default,
+        EntityData<T>: EntityBuilder,
+    {
+        Repo::clone(self, url)
+    }
+
+    // Utils
+
+    pub(super) fn lock(&self) -> MutexGuard<git2::Repository> {
+        self.backend.lock().unwrap()
+    }
+
+    pub(crate) fn has_commit(&self, urn: &RadUrn, oid: git2::Oid) -> Result<bool, Error> {
+        let span = tracing::warn_span!("Storage::has_commit", urn = %urn, oid = %oid);
+        let _guard = span.enter();
+
+        if oid.is_zero() {
+            return Ok(false);
+        }
+
+        let git = self.lock();
+        let commit = git.find_commit(oid);
```

Comment:
Stupid readability nitpick: why not directly `match git.find_commit(oid)`?

Reply:
`borrowck` tells me that `git may not live long enough` (weirdly, because `git` is let-bound), and suggest to `let x = match ...; x` :/ Not sure which is more readable...

Reply:
Nah, this assignment is simpler in the end.